ci/get_commit_message.py

```diff
@@ -73,28 +73,40 @@ def modify_commit_message_pragmas(commit_message, target):
     modified_files = git_files_changed(target)
 
     rec_tags = tags.files_to_tags(modified_files)
+    commit_tags = set()
 
     # Extract all "Features" and "Test-tag"
     feature_tags = re.findall(
         r'^Features:(.*)$', commit_message, flags=re.MULTILINE | re.IGNORECASE)
     if feature_tags:
         for _tags in feature_tags:
-            rec_tags.update(filter(None, _tags.split(' ')))
+            commit_tags.update(filter(None, _tags.split(' ')))
         commit_message = re.sub(
             r'^Features:.*$', '', commit_message, flags=re.MULTILINE | re.IGNORECASE)
     test_tags = re.findall(
         r'^Test-tag:(.*)$', commit_message, flags=re.MULTILINE | re.IGNORECASE)
     if test_tags:
         for _tags in test_tags:
-            rec_tags.update(filter(None, _tags.split(' ')))
+            commit_tags.update(filter(None, _tags.split(' ')))
         commit_message = re.sub(
             r'^Test-tag:.*$', '', commit_message, flags=re.MULTILINE | re.IGNORECASE)
 
     # Put "Test-tag" after the title
     commit_message_split = commit_message.splitlines()
     commit_message_split.insert(1, '')
-    commit_message_split.insert(2, '# Auto-recommended Test-tag')
-    commit_message_split.insert(3, f'Test-tag: {" ".join(sorted(rec_tags))}')
+    if test_tags:
+        # Keep the commit tags but leave a comment with the recommended
+        commit_message_split.insert(2, f'Test-tag: {" ".join(sorted(commit_tags))}')
+        commit_message_split.insert(3, '# Auto-recommended Test-tag')
+        commit_message_split.insert(4, f'# Test-tag: {" ".join(sorted(rec_tags))}')
+        commit_message_split.insert(5, '')
+    else:
+        # If there were "Feature" tags, merge into the recommended.
+        # Otherwise, just use the recommended.
+        rec_tags.update(commit_tags)
+        commit_message_split.insert(2, '# Auto-recommended Test-tag')
+        commit_message_split.insert(3, f'Test-tag: {" ".join(sorted(rec_tags))}')
+        commit_message_split.insert(4, '')
```

Comment:
Still need to shake out the details but the idea is
1. Features and Test-tag are not both supported. If the commit has both, merge as a convenience. I've seen people try to specify both.
2. If the commit has Test-tag, just leave a comment with the recommended. This allows people to override for whatever reason. E.g. when trying iterating on a broken test and not wanting to run everything for every commit.

The end result is that the modified commit never has Features. Only Test-tag

Reply:
When you say "supported" this is all our code I assume, there's some logic that we own in launch.py that extracts this and passes it to the avocado command line?

I'm not saying this is wrong or requesting changes but just trying to understand the requirements and limitations.

Reply:
No. pipeline-lib extracts commit pragmas and converts the Features and Test-tag to a single line that gets passed to launch.py

Reply:
Well, pipeline-lib is also within our control but harder to modify than here certainly.

Reply:
Why is it harder to modify?  It even tests the modifications being made with all currently supported branches to try to prevent unexpected regressions.

Reply:
It's a separate repo, so I'd have to first update pipeline lib and verify it's working and backwards compatible with other branches. And I would still need a daos PR to verify expected behavior of passing both pragmas. Whereas if I modify the daos repo, it's a single atomic PR to make + test the change. And I don't have to worry about breaking some other branch in the process

Reply:
> It's a separate repo, so I'd have to first update pipeline lib and verify it's working

Yes.  Always have to verify changes being made work as intended.  But this isn't really any different than if it were on the same project.

> and backwards compatible with other branches.

`pipeline-lib`'s CI does that for you.  It tests the `pipeline-lib` against daos' `master`, `release/2.4`, `weekly-testing` and `weekly-2.4-testing` branches.  Might want to think about adding the new `daily*` branches in there but just those 4 branches is pretty good coverage.  You can even define a `Test-tag:` you want to be used during all of that testing.

> And I would still need a daos PR to verify expected behavior of passing both pragmas.

Again, that sounds like it's just a normal process of testing/validating one's changes, regardless of which repo they are in.

Anyway, just hoping to clear up any misconceptions on the perceived difficulty of updating `pipeline-lib`.

Reply:
> Yes. Always have to verify changes being made work as intended. But this isn't really any different than if it were on the same project.

1. By using just the daos repo, I can run and debug some code locally to verify it's working. Testing pipeline-lib code locally is virtually impossible.
2. I would need TWO PRS - one in pipeline-lib, and then one in daos that uses the pipeline-lib one. And every time I update the pipeline-lib PR, I would have to retrigger my daos PR.

> > And I would still need a daos PR to verify expected behavior of passing both pragmas.
>
> Again, that sounds like it's just a normal process of testing/validating one's changes, regardless of which repo they are in.

I disagree. If I have a daos PR, I only need ONE PR, and I only need to verify my changes on that PR for that ONE branch. With pipeline-lib, I would have to verify all supported branches. Sure, pipeline-lib has automated steps for some of this, but it's still more complex that just ONE daos PR. And I would still have to manipulate the `Test-tag` and `Features` commit pragmas for those auto-spawned jobs to verify it works. AND I would still need a separate daos PR to verify it works in "PR mode", since pipeline-lib has different behavior depending on whether it's a timed build, PR, upstream, etc. AND it takes much longer to wait for testing from 4 branches than for one branch.

Another benefit to having just a daos PR is that the change is atomic. One PR, one commit - instead of co-dependent PRs in both daos and pipeline-lib.

In all, I find pipeline-lib incredibly cumbersome to work with.

Reply:
> In all, I find pipeline-lib incredibly cumbersome to work with.

Not disagreeing that it's certainly more cumbersome than having everything in a single repo on a single branch.  But then when you do branch you have to start duplicating changes across branches.  The exact reason libraries were invented.

That said, _library_ is not really even our primary reason for `pipeline-lib` (as much as we do benefit from it being a common library to many projects and branches and greatly reducing duplication).  It's primary reason for existence is Jenkins itself and it's limitations.  We've had to move soooo much code out of a daos branch's `Jenkinsfile` into `pipeline-lib` just because there is an upper limit to the size of `Jenkinsfile`.

Reply:
> But then when you do branch you have to start duplicating changes across branches. The exact reason libraries were invented.

But proper libraries are versioned and you can specify which version you want to use. With pipeline-lib, all daos branches are forced to use the same version. Which goes back to the previous problem of having to make sure all active branches  still work.
I don't think putting logic in the daos repo is _**duplicating**_ code - it's just versioned for different branches. We don't want to do the exact same thing for every branch.

Reply:
> there is an upper limit to the size of Jenkinsfile

FWIW I don't think it's the Jenkinsfile that has a limitation on size. It's the `pipeline` object that is limited in size. But either way, this is an external python script and doesn't affect that.